Multiple directories support, norecursedirs, passing args to pytest

aldanor opened the pull request:
I believe this PR closes all outstanding issues (#5, #6, #7 and #8). If merged in, this probably calls for a version bump.

The README will probably need to be updated as well..

Options added / things modified:
- All positional arguments after `--` are now passed as is to `py.test` executable.
- Added support for multiple watched folders (passed as positional arguments).
- Allow to ignore sub folders (currently via an exact match only at a first level of each watched folder, but this could be improved quite easily to support arbitrary patterns). Note that this is done at the observer level and not in the event handler because if you have tens of thousands of files in a subfolder like `.tox` for example, the handler would get completely swamped by that (so a simple pattern matching inside `on_any_event` is not going to cut it).
- py.test subprocess no longer changes directory to a watched folder -- this no longer makes any sense given that any args can be passed to pytest and you can watch multiple folders at once.
- py.test subprocess no longer runs with shell=True, to handle quoted arguments properly.
- Catch `KeyboardInterrupt` even if it was caught after the first test run and exit gracefully.
- Instead of catching all file events, only react to `FileModifiedEvent` and `FileCreatedEvent`. This fixed the problem where duplicate watchdog events would be fired upon a single file modification if using remote NFS mounts on Linux.
- Added `-p` alias to `--poll`.
- The exact py.test command is now printed to stdout upon the first test run.
- Added event spooling with 200ms cooldown to work around atomic saves in different editors.
- Added `--verbose` / `--quiet` options.
- Added `--no-spool` option to disable spooling.
- `--ext` option will prepend extensions with periods if needed.

An example using the new functionality:
```sh
ptw -v -p --no-spool --ignore=.tox,.git folder1 folder2 -- tests/*.py -svv
```

joeyespo commented:
This all looks great! Thank you for putting this all together, @aldanor!

Also, really clear PR message. I really appreciate the writeup. It set good expectations. And the example usage was helpful to envision these changes.

I'm all ready to pull it in. Just address those few comments first. For convenience:

- [ ] Use `[directories]` instead of `--watch` to keep it simple and backwards compatible?
- [ ] Can you think of any cases where `shell=True` is an advantage here?
- [ ] If not, go ahead and remove the redundant `shell=False`
- [ ] Are file renames picked up by the current events?
- [ ] `--ignore` instead of `--nonrecursedirs`?

I can also bump the version and update the README after.

joeyespo commented:
Nevermind about `--norecursedirs`. I just remembered that came from py.test.

Do you think we should also pass that argument to py.test? Otherwise, py.test might still run tests even though we're not listening for changes in those deeper directories.

aldanor commented:
Renaming `--norecursedirs` to `--ignore` seems reasonable. Passing it to `py.test` automatically is probably not, as you may want to listen to a restricted set of subfolders while still running the full test suite, I think -- these are two different notions. You could do it the other way around though and use `norecursedirs` from pytest config file (which is most always present anyway) as a default value for `--ignore`.

aldanor commented:
@joeyespo The `--watch` vs `[directory]` (`[directories]`) question is still open -- if you can suggest an unambiguous docopt signature for this I'd be happy to fix it.

aldanor commented:
@joeyespo CLI args fixed, you can try testing it now. What's remaining is event spooling, will hack on it now.

aldanor commented:
@joeyespo Event spooling done:

```
Changes detected:
    deleted:  pytest_watch/watcher.py
    moved:    pytest_watch/.sublc49.tmp -> pytest_watch/watcher.py

Rerunning pytest command: py.test --import .
```

I guess I'll have to add `-v` / `-q` options for verbose/quiet output, and also an option to opt out of event spooling (should it be enabled by default? the default cooldown is 200ms).

blueyed commented:
@aldanor 
I've noticed that the text of the modified filenames and the command in "Rerunning pytest command" is not visible in my terminal, but can be copy'n'pasted.

blueyed commented:
Instead of using `highlight = lambda arg: Fore.LIGHTWHITE_EX + arg + Fore.CYAN` it might be better to just make it bold?

tarpas commented:
On Mac OS the spooling seems to cause problems.
tibor$ ptw

Running pytest command: py.test
=============================================== test session starts ================================================
platform darwin -- Python 2.7.8 -- py-1.4.28 -- pytest-2.7.1
rootdir: /Users/tibor/tmonworkspace/testmon/exampleproject, inifile: pytest.ini
plugins: testmon, cache, html
collected 13 items 
  tests/test_a.py sx..  tests/test_ab.py .  tests/test_b.py ..  tests/single/test_s.py ...  tests/single/test_t.py ..  tests_2/test_e.py s
================================= 10 passed, 2 skipped, 1 xfailed in 0.08 seconds ==================================
Process Timer-1:
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/Users/tibor/tmonworkspace/pytest-watch/pytest_watch/spooler.py", line 21, in run
    self.function(*self.args, **self.kwargs)
  File "/Users/tibor/tmonworkspace/pytest-watch/pytest_watch/spooler.py", line 39, in process
    self.callback([self.outbox.get() for _ in range(self.outbox.qsize())])
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/multiprocessing/queues.py", line 143, in qsize
    return self._maxsize - self._sem._semlock._get_value()
NotImplementedError

joeyespo commented:
@aldanor There's a lot of divergent subtle cases being handled in this PR. If it's getting too unwieldy, perhaps open a few different ones to address specific issues. I'm very ready to pull in most things in here, but until it's completely ready, the whole thing is held up.

aldanor commented:
I got to my mac, will run the whole thing on it tonight.

I'm actually thinking this whole thing needs a test suite...

aldanor commented:
@joeyespo Yep I know.... it will be real hard for me to start going backwards and split it up because I have to use it on daily basis, which means I'll have to keep track of more than three versions now -- remote, local and multiple PR branches... :(

I'd rather make this whole thing work. (the norecursedirs changes I can roll back -- that's the only truly questionable part which will need further effort and thought).

aldanor commented:
@tarpas Fixed. OSX weirdness...

aldanor commented:
@blueyed Fixed

aldanor commented:
Anything else?

tarpas commented:
Works without problems for me also with pytest testmon. Thanks for the
work!  @joyespo it would be great if you merge it.

On Wed, Jun 10, 2015 at 11:02 PM, Joe Esposito <notifications@github.com>
wrote:

> @aldanor <https://github.com/aldanor> There's a lot of divergent subtle
> cases being handled in this PR. If it's getting too unwieldy, perhaps open
> a few different ones to address specific issues. I'm very ready to pull in
> most things in here, but until it's completely ready, the whole thing is
> held up.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/joeyespo/pytest-watch/pull/11#issuecomment-110912714>.
>

joeyespo commented:
> Yep I know.... it will be real hard for me to start going backwards and split it up because I have to use it on daily basis, which means I'll have to keep track of more than three versions now -- remote, local and multiple PR branches

Yeah, it'd be kind of tough now. Oh well.

Pulling this is now so we can move forward. We can fix up `--ignore` / `--norecursedirs` stuff after. Lots of good work in this PR.

Many thanks, @aldanor :smiley:

joeyespo commented:
@blueyed Did 0a243761465e2e4087fdf007ce533b16c70ab2dd do what you need? If not, could you open a new issue?

joeyespo commented:
Thanks again, @aldanor for your fantastic work here. And for your patience :-)

aldanor commented:
@joeyespo Np :) I think all the outstanding issues have been covered, everyone's happy? (`--ignore` is a separate issue but it actually does somewhat work already)

Btw, maybe it would make sense to move this thing over to http://github.com/pytest-dev? (seeing as pytest itself will move there soon)

aldanor commented:
This one was one hell of a PR, more Python LOC than the original code :)

joeyespo commented:
> Btw, maybe it would make sense to move this thing over to http://github.com/pytest-dev?

I'm up for it if they are.

> This one was one hell of a PR, more Python LOC than the original code :)

:smiley:

aldanor commented:
What exactly is Linux-only? I'm mostly on OSX myself (occasionally on Linux)

joeyespo commented:
@aldanor Oops, meant Unix, not Linux. [`os.path.samefile`](https://docs.python.org/2/library/os.path.html#os.path.samefile)

blueyed commented:
@joeyespo 
> @blueyed Did 0a24376 do what you need? If not, could you open a new issue?

Yes, thanks!

joeyespo commented:
FYI released these changes today. Be sure to `pip install --upgrade pytest-watch` :smiley:
